### PR TITLE
Classed refactor

### LINKED ISSUE #2219: [Component] classed() endpoint
The `classed` endpoint. [jQuery](https://api.jquery.com/addclass/) [is](https://api.jquery.com/hasclass/) [nicer](https://api.jquery.com/removeclass/)

Our current implementation:
```typescript
    /**
     * Checks if the Component has a given CSS class.
     *
     * @param {string} cssClass The CSS class to check for.
     */
    public classed(cssClass: string): boolean;
    /**
     * Adds/removes a given CSS class to/from the Component.
     *
     * @param {string} cssClass The CSS class to add or remove.
     * @param {boolean} addClass If true, adds the provided CSS class; otherwise, removes it.
     * @returns {Component} The calling Component.
     */
    public classed(cssClass: string, addClass: boolean): Component;
```

Pros: 
- Consistent with `d3.classed()`

Cons:
- It is ugly
- It has a boolean as a parameter

### PR DESCRIPTION
**Pull request by aicioara**
Closes #2219 

> Currently the CSS classes on a Component are queried and set using the classed() method:
> 
> component.classed("foo"); // does the Component have the class "foo"?
> component.classed("foo", true); // adds the class "foo" to the Component
> component.classed("foo", false); // removes the class "foo" from the Component
> 
> This mirrors the D3 method of the same name (although with more limited functionality). However, the boolean parameter leads to mistakes because it effectively functions as a trinary parameter (true/false/undefined to switch between "add"/"remove"/"query"):
> 
> component.classed("foo"); // sounds like it might add the class "foo", but doesn't
> 
> Our proposed alternative is to use separate methods for each part of the functionality:
> 
> component.hasClass("foo");
> component.addClass("foo");
> component.removeClass("foo");
> 
> While this is less consistent with D3, we think it will cut down on confusion. Thoughts?

### CODE REVIEW

**Comment by aicioara** [comment]
addClass RegEx
```regex
\.classed\([^,)]*, true\)
replace with

```

**Comment by aicioara** [comment]
Not sure what to say about change anymore. `.classed().` is still used in all these places

```typescript
         this._element.classed(cssClass, true);
       this._backgroundContainer = this._element.append("g").classed("background-container", true);
       this._content = this._element.append("g").classed("content", true);
       this._foregroundContainer = this._element.append("g").classed("foreground-container", true);
       this._boxContainer = this._element.append("g").classed("box-container", true);
       if (className != null) { box.classed(className, true); }
         return this._element.classed(cssClass);
         this._element.classed(cssClass, true);
         this._element.classed(cssClass, false);
       this._detectionEdgeT = createLine().classed("drag-edge-tb", true);
       this._detectionEdgeB = createLine().classed("drag-edge-tb", true);
       this._detectionEdgeL = createLine().classed("drag-edge-lr", true);
       this._detectionEdgeR = createLine().classed("drag-edge-lr", true);
         this._detectionCornerTL = createCorner().classed("drag-corner-tl", true);
         this._detectionCornerTR = createCorner().classed("drag-corner-tr", true);
         this._detectionCornerBL = createCorner().classed("drag-corner-bl", true);
         this._detectionCornerBR = createCorner().classed("drag-corner-br", true);
       this._xLinesContainer = this.content().append("g").classed("x-gridlines", true);
       this._yLinesContainer = this.content().append("g").classed("y-gridlines", true);
              .classed("zeroline", (t: number) => t === 0);
               .classed("zeroline", (t: number) => t === 0);
       this._swatchContainer = this.content().append("g").classed("swatch-container", true);
       this._swatchBoundingBox = this.content().append("rect").classed("swatch-bounding-box", true);
       this._lowerLabel = this.content().append("g").classed(InterpolatedColorLegend.LEGEND_LABEL_CLASS, true);
       this._upperLabel = this.content().append("g").classed(InterpolatedColorLegend.LEGEND_LABEL_CLASS, true);
       swatches.enter().append("rect").classed("swatch", true);
       var fakeLegendRow = this.content().append("g").classed(Legend.LEGEND_ROW_CLASS, true);
       var fakeLegendEntry = fakeLegendRow.append("g").classed(Legend.LEGEND_ENTRY_CLASS, true);
       rows.enter().append("g").classed(Legend.LEGEND_ROW_CLASS, true);
       var entriesEnter = entries.enter().append("g").classed(Legend.LEGEND_ENTRY_CLASS, true);
       entriesEnter.append("g").classed("text-container", true);
                             .classed(Legend.LEGEND_SYMBOL_CLASS, true);
       this._box = this.content().append("g").classed("selection-box", true).remove();
       this._boxArea = this._box.append("rect").classed("selection-area", true);
                                             .classed(Axis.TICK_MARK_CLASS + "-container", true);
                                              .classed(Axis.TICK_LABEL_CLASS + "-container", true);
       this._baseline = this.content().append("line").classed("baseline", true);
       tickMarks.enter().append("line").classed(Axis.TICK_MARK_CLASS, true);
       d3.select(tickMarks[0][0]).classed(Axis.END_TICK_MARK_CLASS, true)
       d3.select(tickMarks[0][tickMarkValues.length - 1]).classed(Axis.END_TICK_MARK_CLASS, true)
       tickLabels.enter().append("g").classed(Axis.TICK_LABEL_CLASS, true);
       tickLabels.enter().append("text").classed(Axis.TICK_LABEL_CLASS, true);
         var tierContainer = this.content().append("g").classed(Time.TIME_AXIS_TIER_CLASS, true);
         this._tierLabelContainers.push(tierContainer.append("g").classed(Axis.TICK_LABEL_CLASS + "-container", true));
         this._tierMarkContainers.push(tierContainer.append("g").classed(Axis.TICK_MARK_CLASS + "-container", true));
         this._tierBaselines.push(tierContainer.append("line").classed("baseline", true));
       var tickLabelsEnter = tickLabels.enter().append("g").classed(Axis.TICK_LABEL_CLASS, true);
       tickMarks.enter().append("line").classed(Axis.TICK_MARK_CLASS, true);
       d3.select(tickMarks[0][0]).classed(Axis.END_TICK_MARK_CLASS, true);
       d3.select(tickMarks[0][tickMarks.size() - 1]).classed(Axis.END_TICK_MARK_CLASS, true);
       tickMarks.enter().append("line").classed(Axis.TICK_MARK_CLASS, true);
       this._baseline = this._renderArea.append("line").classed("baseline", true);
       drawer.renderArea().classed(Bar._BAR_AREA_CLASS, true);
       var labelArea = this._renderArea.append("g").classed(Bar._LABEL_AREA_CLASS, true);
           g.classed(className, true);
       this._renderArea = this.content().append("g").classed("render-area", true);
       this._baseline = this._renderArea.append("line").classed("baseline", true);
       this._areaSelection.classed(Area.PATH_CLASS, true);
         dataElements.classed(this._className, true);
                                 .classed(Line.PATH_CLASS, true)
       this._pathSelection.classed(Line.PATH_CLASS, true);
       colorTester.classed(className, true);
       colorTester.classed(className, false);
     assert.isTrue((<any> label)._element.classed("label"), "title element has label css class");
     assert.isTrue((<any> label)._element.classed("title-label"), "title element has title-label css class");
     assert.isTrue(firstTick.classed(Plottable.Axis.END_TICK_MARK_CLASS), "first end tick has the end-tick-mark class");
     assert.isTrue(lastTick.classed(Plottable.Axis.END_TICK_MARK_CLASS), "last end tick has the end-tick-mark class");
       assert.isTrue(areaPath.classed("pink"));
       assert.isTrue(areaPath.classed(Plottable.Drawers.Area.PATH_CLASS));
       assert.isTrue(linePath.classed("pink"));
       assert.isTrue(linePath.classed(Plottable.Drawers.Line.PATH_CLASS));
       assert.isTrue(svg.classed("plottable"), "<svg> was given \"plottable\" CSS class");
       assert.isTrue(svg2.classed("plottable"), "second <svg> was given \"plottable\" CSS class");
     assert.isTrue(g0.classed("background-container"), "the first g is a background container");
     assert.isTrue(g1.classed("content"), "the second g is a content container");
     assert.isTrue(g2.classed("foreground-container"), "the third g is a foreground container");
     assert.isTrue(g3.classed("box-container"), "the fourth g is a box container");
     var colorTester = d3.select("body").append("div").classed("color-tester", true);

```

**Comment by jtlan** [comment]
@aicioara wrote:
>Not sure what to say about change anymore. `.classed().` is still used in all these places
<snip>

Yeah, but those are all D3 calls, so nothing we can do about that.

**Comment by aicioara** [comment]
What I mean is that we're now inconsistent

**Comment by jtlan** [comment]
Can we use a `Utils.Set` to keep track of what classes we have?

File: test/core/componentTests.ts [hunk 1]
```diff
@@ -298,24 +298,24 @@ describe("Component behavior", () => {
...
-    assert.strictEqual(c.classed(undefined, true), c, "returns this when classed called w/ undefined and true");
+    assert.isTrue(c.hasClass("CSS-PREANCHOR-KEEP"));
+    assert.isFalse(c.hasClass("CSS-PREANCHOR-REMOVE"));
+    assert.isFalse(c.hasClass("CSS-POSTANCHOR"));
+    c.addClass("CSS-POSTANCHOR");
+    assert.isTrue(c.hasClass("CSS-POSTANCHOR"));
+    c.removeClass("CSS-POSTANCHOR");
+    assert.isFalse(c.hasClass("CSS-POSTANCHOR"));
+    assert.isFalse(c.hasClass(undefined), "returns false when classed called w/ undefined");
+    assert.strictEqual(c.addClass(undefined), c, "returns this when classed called w/ undefined and true");
```
Comment (jtlan) [inline test/core/componentTests.ts]: Probably want to change the message on this line and the line above.

File: src/components/component.ts [hunk 2]
```diff
@@ -263,8 +267,8 @@ module Plottable {
         this.anchor(selection);
       }
       if (this._element == null) {
-        throw new Error("If a Component has never been rendered before, then renderTo must be given a node to render to, \
-          or a d3.Selection, or a selector string");
+        throw new Error("If a Component has never been rendered before, then renderTo must be given a node to render to, " +
+          "or a d3.Selection, or a selector string");
```
Comment (jtlan) [inline src/components/component.ts]: Why this change?

Comment (aicioara) [inline src/components/component.ts]: TBH, main reason was that my Sublime parser was dying here. However, this is better practice.

I generally dislike trailing `\` because misconfigurations in the editor might add an invisible whitespace right after the `\` and lead to errors. This format is more robust

**Comment by aicioara** [comment]
Restarted the build. SauceLabs failed.

**Comment by jtlan** [comment]
[+1]

QE, probably something like a regression pass.

**Comment by aicioara** [comment]
In particular CSS classes should be applied to all components nicely (axes, plots).

**Comment by bluong** [comment]
This is actually a change in functionality, so a proper QE pass I think is needed.

Adding classes, removing classes, and checking for classes should hopefully have a smoother usage.

**Comment by crmorford** [comment]
for myself: http://jsfiddle.net/3y4ty91q/
